mempool/mempool_test.go

```diff
@@ -866,3 +930,851 @@ func TestCheckSpend(t *testing.T) {
+				// spends it, but also replaces it, should be
+				// invalid.
+				parentOut := txOutToSpendableOut(parent, 0)
+				outs = []spendableOutput{coinbaseOut, parentOut}
+				tx, err := ctx.harness.CreateSignedTx(
+					outs, 2, defaultFee, false,
+				)
+				if err != nil {
+					ctx.t.Fatalf("unable to create "+
+						"transaction: %v", err)
+				}
+
+				return tx, nil
+			},
+			err: "spends parent transaction",
+		},
+		{
+			// A transaction cannot replace another if it has a
+			// lower fee rate than any of the transactions it
+			// intends to replace.
+			name: "insufficient fee rate",
+			setup: func(ctx *testContext) (*btcutil.Tx, []*btcutil.Tx) {
+				coinbase1 := ctx.addCoinbaseTx(1)
+				coinbase2 := ctx.addCoinbaseTx(1)
+
+				// We'll create two transactions that each spend
+				// one of the coinbase outputs. The first will
+				// have a higher fee rate than the second.
+				coinbaseOut1 := txOutToSpendableOut(coinbase1, 0)
+				outs := []spendableOutput{coinbaseOut1}
+				ctx.addSignedTx(outs, 1, defaultFee*2, true, false)
+
+				coinbaseOut2 := txOutToSpendableOut(coinbase2, 0)
+				outs = []spendableOutput{coinbaseOut2}
+				ctx.addSignedTx(outs, 1, defaultFee, true, false)
+
+				// We'll then create the replacement transaction
+				// by spending the coinbase outputs. It will be
+				// an invalid one however, since it won't have a
+				// higher fee rate than the first transaction.
+				outs = []spendableOutput{coinbaseOut1, coinbaseOut2}
+				tx, err := ctx.harness.CreateSignedTx(
+					outs, 1, defaultFee*2, false,
+				)
+				if err != nil {
+					ctx.t.Fatalf("unable to create "+
+						"transaction: %v", err)
+				}
+
+				return tx, nil
+			},
+			err: "insufficient fee rate",
+		},
+		{
+			// A transaction cannot replace another if it doesn't
+			// have an absolute greater than the transactions its
+			// replacing _plus_ the replacement transaction's
+			// minimum relay fee.
+			name: "insufficient absolute fee",
+			setup: func(ctx *testContext) (*btcutil.Tx, []*btcutil.Tx) {
+				coinbase := ctx.addCoinbaseTx(1)
+
+				// We'll create a transaction with two outputs
+				// and the default fee.
+				coinbaseOut := txOutToSpendableOut(coinbase, 0)
+				outs := []spendableOutput{coinbaseOut}
+				ctx.addSignedTx(outs, 2, defaultFee, true, false)
+
+				// We'll create a replacement transaction with
+				// one output, which should cause the
+				// transaction's absolute fee to be lower than
+				// the above's, so it'll be invalid.
+				tx, err := ctx.harness.CreateSignedTx(
+					outs, 1, defaultFee, false,
+				)
+				if err != nil {
+					ctx.t.Fatalf("unable to create "+
+						"transaction: %v", err)
+				}
+
+				return tx, nil
+			},
+			err: "insufficient absolute fee",
+		},
+		{
+			// A transaction cannot replace another if it introduces
+			// a new unconfirmed input that was not already in any
+			// of the transactions it's directly replacing.
+			name: "spends new unconfirmed input",
```

Comment:
This is interesting... Any idea what this rule achieves?

Reply:
https://github.com/bitcoin/bitcoin/blob/03858b23fe1d17e681eb18d24602b276020295a0/src/validation.cpp#L709